README.md
Windows Schannel Cookbook
=========================

Description
-------------
A cookbook to configure the windows Secure Channel [(Schannel)](https://msdn.microsoft.com/en-us/library/windows/desktop/aa380123) security support provider (SSP).
This SSP contains a set of security protocols and ciphers suites commonly used for HTTPS communications.  Many of the microsoft provided services and 3rd party software make use of Schannel for their communication so by removing support for a given protocol, removes it for all components.

This can be a double-edge sword, restricting a protocol might be fine for one service and end up breaking another.

Always make sure you __test__ these settings __before__ deploying into __production__ as you could end up __shooting yourself in the foot__

Using this cookbook should help improve your security posture for the following microsoft services and help towards meeting regulatory compliance (PCI CIS FIPS) 

* Internet Information Services (IIS)
* Microsoft Remote Desktop (RDP) _when configured to use TLS_
* LDAPS (Active Directory)
* Global Catalog SSL (Active Directory)
* WinRM/HTTPS (Windows Remote Management)

TODO : Insert image of external scan before and after.

Support for managing the following
----------------
### Protocols
+ Transport Layer Security (TLS) - v1.0, v1.1, v1.2
+ Secure Socket Layer (SSL) - v2.0, v3.0
+ Private Communications Technology (PCT) - v1.0

### Cipher Suites
+ NULL, DES, RC2, RC4, 3DES, AES

### Hashes
+ MD5, SHA

### Key Exchanges
+ Diffie-Hellman (DH)

Requirements
------------
#### Platforms
* Windows Server 2008 (R1, R2)
* Windows Server 2012 (R1, R2)

Attributes
----------
* `node['windows_schannel']['allow_reboot']` = `false` 
_changes made to protocols won't take affect until after a restart, this doesn't apply to ciphers and hashes. The default is false change this to true to initiate a reboot after the run._

Other than the `allow_reboot` attribute all the others are controlled by either setting it to either `enable` or `disable`
_this is CaSe Sensitive_

#### Protocols
* `node['windows_schannel']['pct_v1.0']` = `disable`
* `node['windows_schannel']['ssl_v2.0']` = `disable`
* `node['windows_schannel']['ssl_v3.0']` = `disable`
* `node['windows_schannel']['tls_v1.0']` = `enable`
* `node['windows_schannel']['tls_v1.1']` = `enable`
* `node['windows_schannel']['tls_v1.2']` = `enable`

#### Ciphers
* `node['windows_schannel']['cipher_null']` = `disable`
* `node['windows_schannel']['cipher_des']` = `disable`
* `node['windows_schannel']['cipher_rc2']` = `disable`
* `node['windows_schannel']['cipher_rc4']` = `disable`
* `node['windows_schannel']['cipher_3des']` = `enable`
* `node['windows_schannel']['cipher_aes']` = `enable`

#### Hashes
* `node['windows_schannel']['hash_md5']` = `enable`
* `node['windows_schannel']['hash_sha']` = `enable`

#### Key Exchanges
* `node['windows_schannel']['keyEx_dh']` = `enable`

License & Authors
-----------------
- Author:: Gary Bright (<digitalgaz@hotmail.com>)

```text
Copyright 2017, Gary Bright

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
```

CHANGELOG.md
# windows_schannel Cookbook CHANGELOG

This file is used to list changes made in each version of the windows_schannel cookbook.

# 1.0.0 - (06-10-2017)

- Bumping Cookbook to a big 1.0.0.  It's been used in production for over a year and I think it's in a strong shape, all functionatly tested.

# 0.4.1 - (06-10-2017)

- The registry key for the 3DES Cipher was incorrect and as a result wasn't managing it correctly

# 0.4.0 - (06-10-2017)

- Migrated the integration tests to inspec.
- Changed some of the wording of the ChefSpec Tests

## 0.3.0 - (06-10-2017)

- Corrected ChefSpec errors, warnings and now has 100% coverage

## 0.2.0 - (06-10-2017)

- Refreshed cookbook metadata, corrected unit & chef style linting errors

## 0.1.4 - (19-09-2016)

- Added RC4 128/128 to the RC4 cipher suite with RC4 128/128 unit test (Thank You @brianbohanon)

## 0.1.3 - (20-06-2016)

- Changed the description to make it easyer to understand what it does in the supermarket

## 0.1.2 - (17-06-2016)

- Fixed some typo's in the README ready for relase to the supermarket.

## v0.1.1 (16-06-2016)

- Changed the attribute values to make it easier to read rather than 1 or 0, means it is now caSe sensitive
but I think its worth the trade off as its easier to interpret
- Added support for Hashes (MD5 and SHA) and Keyexchanges (DH)
- Refactored the ciphers to loop based on foodcritic
- Added reboot notifier for protocol changes.

## v0.1.0 (20-04-2016)

- Building the framework of the cookbook.


spec/unit/recipes/hashes_spec.rb
#
# Cookbook Name:: windows_schannel
# Spec:: default
#
# Copyright (c) 2016 The Authors, All Rights Reserved.

require 'spec_helper'

describe 'windows_schannel::hashes' do
  context 'When all attributes are default, on an windows platform it' do
    let(:chef_run) do
      runner = ChefSpec::ServerRunner.new(platform: 'windows', version: '2012r2')
      runner.converge(described_recipe)
    end
    it 'configures the use of the MD5 Hash' do
      expect(chef_run).to create_registry_key('hash_md5')
    end
    it 'configures the use of the SHA Hash' do
      expect(chef_run).to create_registry_key('hash_sha')
    end
    it 'converges successfully' do
      expect { chef_run }.to_not raise_error
    end
  end
end


spec/unit/recipes/ciphers_spec.rb
#
# Cookbook Name:: windows_schannel
# Spec:: default
#
# Copyright (c) 2016 The Authors, All Rights Reserved.

require 'spec_helper'

describe 'windows_schannel::ciphers' do
  context 'When all attributes are default, on an windows platform it' do
    let(:chef_run) do
      runner = ChefSpec::ServerRunner.new(platform: 'windows', version: '2012r2')
      runner.converge(described_recipe)
    end
    it 'configures the cipher NULL' do
      expect(chef_run).to create_registry_key('cipher_null')
    end
    it 'configures the cipher DES' do
      expect(chef_run).to create_registry_key('cipher_des')
    end
    it 'configures the cipher RC2' do
      expect(chef_run).to create_registry_key('RC2 40/128')
      expect(chef_run).to create_registry_key('RC2 56/128')
      expect(chef_run).to create_registry_key('RC2 128/128')
    end
    it 'configures the cipher RC4' do
      expect(chef_run).to create_registry_key('RC4 40/128')
      expect(chef_run).to create_registry_key('RC4 56/128')
      expect(chef_run).to create_registry_key('RC4 64/128')
      expect(chef_run).to create_registry_key('RC4 128/128')
    end
    it 'configures the cipher 3DES' do
      expect(chef_run).to create_registry_key('cipher_3des')
    end
    it 'configures the cipher AES' do
      expect(chef_run).to create_registry_key('AES 128/128')
      expect(chef_run).to create_registry_key('AES 256/256')
    end
    it 'converges successfully' do
      expect { chef_run }.to_not raise_error
    end
  end
end


spec/unit/recipes/default_spec.rb
#
# Cookbook Name:: windows_schannel
# Spec:: default
#
# Copyright (c) 2016 The Authors, All Rights Reserved.

require 'spec_helper'

describe 'windows_schannel::default' do
  context 'When all attributes are default, on an windows platform it' do
    let(:chef_run) do
      runner = ChefSpec::ServerRunner.new(platform: 'windows', version: '2012r2')
      runner.converge(described_recipe)
    end

    it 'converges successfully' do
      expect { chef_run }.to_not raise_error
    end
  end
end


recipes/hashes.rb
#
# Cookbook Name:: windows_schannel
# Recipe:: hashes
#
# Copyright (c) 2016 The Authors, All Rights Reserved.
case node['platform']
when 'windows'
  if node['platform_version'].to_f >= 6.0 || node['platform_version'].to_f <= 6.3
    registry_key 'hash_md5' do
      key 'HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Control\SecurityProviders\SCHANNEL\Hashes\MD5'
      recursive true
      case node['windows_schannel']['hash_md5']
      when 'disable'
        values(name: 'Enabled', type: :dword, data: 0)
        :create
      when 'enable'
        values(name: 'Enabled', type: :dword, data: 4_294_967_295)
        :create
      end
    end
    registry_key 'hash_sha' do
      key 'HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Control\SecurityProviders\SCHANNEL\Hashes\SHA'
      recursive true
      case node['windows_schannel']['hash_sha']
      when 'disable'
        values(name: 'Enabled', type: :dword, data: 0)
        :create
      when 'enable'
        values(name: 'Enabled', type: :dword, data: 4_294_967_295)
        :create
      end
    end
  end
end


recipes/protocols.rb
#
# Cookbook Name:: windows_schannel
# Recipe:: protocols
#
# Copyright (c) 2016 The Authors, All Rights Reserved.
case node['platform']
when 'windows'
  if node['platform_version'].to_f >= 6.0 || node['platform_version'].to_f <= 6.3

    reboot 'computer' do
      action :nothing
      delay_mins 1
      reason 'chef - windows_schannel - Changes made Schannel Protocols requires a reboot to take affect.'
      only_if { node['windows_schannel']['allow_reboot'] == true }
    end

    registry_key 'pct_v1.0' do
      key 'HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Control\SecurityProviders\SCHANNEL\Protocols\PCT 1.0\Server'
      recursive true
      case node['windows_schannel']['pct_v1.0']
      when 'disable'
        values [{ name: 'Enabled', type: :dword, data: 0 },
                { name: 'DisabledByDefault', type: :dword, data: 1 }]
        :create
      when 'enable'
        values [{ name: 'Enabled', type: :dword, data: 4_294_967_295 },
                { name: 'DisabledByDefault', type: :dword, data: 0 }]
        :create
      end
      notifies :request_reboot, 'reboot[computer]'
    end

    registry_key 'ssl_v2.0' do
      key 'HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Control\SecurityProviders\SCHANNEL\Protocols\SSL 2.0\Server'
      recursive true
      case node['windows_schannel']['ssl_v2.0']
      when 'disable'
        values [{ name: 'Enabled', type: :dword, data: 0 },
                { name: 'DisabledByDefault', type: :dword, data: 1 }]
        :create
      when 'enable'
        values [{ name: 'Enabled', type: :dword, data: 4_294_967_295 },
                { name: 'DisabledByDefault', type: :dword, data: 0 }]
        :create
      end
      notifies :request_reboot, 'reboot[computer]'
    end

    registry_key 'ssl_v3.0' do
      key 'HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Control\SecurityProviders\SCHANNEL\Protocols\SSL 3.0\Server'
      recursive true
      case node['windows_schannel']['ssl_v3.0']
      when 'disable'
        values [{ name: 'Enabled', type: :dword, data: 0 },
                { name: 'DisabledByDefault', type: :dword, data: 1 }]
        :create
      when 'enable'
        values [{ name: 'Enabled', type: :dword, data: 4_294_967_295 },
                { name: 'DisabledByDefault', type: :dword, data: 0 }]
        :create
      end
      notifies :request_reboot, 'reboot[computer]'
    end

    registry_key 'tls_v1.0' do
      key 'HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Control\SecurityProviders\SCHANNEL\Protocols\TLS 1.0\Server'
      recursive true
      case node['windows_schannel']['tls_v1.0']
      when 'disable'
        values [{ name: 'Enabled', type: :dword, data: 0 },
                { name: 'DisabledByDefault', type: :dword, data: 1 }]
        :create
      when 'enable'
        values [{ name: 'Enabled', type: :dword, data: 4_294_967_295 },
                { name: 'DisabledByDefault', type: :dword, data: 0 }]
        :create
      end
      notifies :request_reboot, 'reboot[computer]'
    end

    registry_key 'tls_v1.1' do
      key 'HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Control\SecurityProviders\SCHANNEL\Protocols\TLS 1.1\Server'
      recursive true
      case node['windows_schannel']['tls_v1.1']
      when 'disable'
        values [{ name: 'Enabled', type: :dword, data: 0 },
                { name: 'DisabledByDefault', type: :dword, data: 1 }]
        :create
      when 'enable'
        values [{ name: 'Enabled', type: :dword, data: 4_294_967_295 },
                { name: 'DisabledByDefault', type: :dword, data: 0 }]
        :create
      end
      notifies :request_reboot, 'reboot[computer]'
    end

    registry_key 'tls_v1.2' do
      key 'HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Control\SecurityProviders\SCHANNEL\Protocols\TLS 1.2\Server'
      recursive true
      case node['windows_schannel']['tls_v1.2']
      when 'disable'
        values [{ name: 'Enabled', type: :dword, data: 0 },
                { name: 'DisabledByDefault', type: :dword, data: 1 }]
        :create
      when 'enable'
        values [{ name: 'Enabled', type: :dword, data: 4_294_967_295 },
                { name: 'DisabledByDefault', type: :dword, data: 0 }]
        :create
      end
      notifies :request_reboot, 'reboot[computer]'
    end
  end
end


recipes/ciphers.rb
#
# Cookbook Name:: windows_schannel
# Recipe:: ciphers
#
# Copyright (c) 2016 The Authors, All Rights Reserved.
case node['platform']
when 'windows'
  if node['platform_version'].to_f >= 6.0 || node['platform_version'].to_f <= 6.3
    registry_key 'cipher_null' do
      key 'HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Control\SecurityProviders\SCHANNEL\Ciphers\NULL'
      recursive true
      case node['windows_schannel']['cipher_null']
      when 'disable'
        values(name: 'Enabled', type: :dword, data: 0)
        :create
      when 'enable'
        values(name: 'Enabled', type: :dword, data: 4_294_967_295)
        :create
      end
    end

    registry_key 'cipher_des' do
      key 'HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Control\SecurityProviders\SCHANNEL\Ciphers\DES 56/56'
      recursive true
      case node['windows_schannel']['cipher_des']
      when 'disable'
        values(name: 'Enabled', type: :dword, data: 0)
        :create
      when 'enable'
        values(name: 'Enabled', type: :dword, data: 4_294_967_295)
        :create
      end
    end

    case node['windows_schannel']['cipher_rc2']
    when 'disable'
      ['RC2 40/128', 'RC2 56/128', 'RC2 128/128'].each do |cipher|
        registry_key cipher do
          key "HKEY_LOCAL_MACHINE\\SYSTEM\\CurrentControlSet\\Control\\SecurityProviders\\SCHANNEL\\Ciphers\\#{cipher}"
          recursive true
          values(name: 'Enabled', type: :dword, data: 0)
          :create
        end
      end
    when 'enable'
      ['RC2 40/128', 'RC2 56/128', 'RC2 128/128'].each do |cipher|
        registry_key cipher do
          key "HKEY_LOCAL_MACHINE\\SYSTEM\\CurrentControlSet\\Control\\SecurityProviders\\SCHANNEL\\Ciphers\\#{cipher}"
          recursive true
          values(name: 'Enabled', type: :dword, data: 4_294_967_295)
          :create
        end
      end
    end

    case node['windows_schannel']['cipher_rc4']
    when 'disable'
      ['RC4 40/128', 'RC4 56/128', 'RC4 64/128', 'RC4 128/128'].each do |cipher|
        registry_key cipher do
          key "HKEY_LOCAL_MACHINE\\SYSTEM\\CurrentControlSet\\Control\\SecurityProviders\\SCHANNEL\\Ciphers\\#{cipher}"
          recursive true
          values(name: 'Enabled', type: :dword, data: 0)
          :create
        end
      end
    when 'enable'
      ['RC4 40/128', 'RC4 56/128', 'RC4 64/128', 'RC4 128/128'].each do |cipher|
        registry_key cipher do
          key "HKEY_LOCAL_MACHINE\\SYSTEM\\CurrentControlSet\\Control\\SecurityProviders\\SCHANNEL\\Ciphers\\#{cipher}"
          recursive true
          values(name: 'Enabled', type: :dword, data: 4_294_967_295)
          :create
        end
      end
    end

    registry_key 'cipher_3des' do
      key 'HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Control\SecurityProviders\SCHANNEL\Ciphers\Triple DES 168'
      recursive true
      case node['windows_schannel']['cipher_3des']
      when 'disable'
        values(name: 'Enabled', type: :dword, data: 0)
        :create
      when 'enable'
        values(name: 'Enabled', type: :dword, data: 4_294_967_295)
        :create
      end
    end

    case node['windows_schannel']['cipher_aes']
    when 'disable'
      ['AES 128/128', 'AES 256/256'].each do |cipher|
        registry_key cipher do
          key "HKEY_LOCAL_MACHINE\\SYSTEM\\CurrentControlSet\\Control\\SecurityProviders\\SCHANNEL\\Ciphers\\#{cipher}"
          recursive true
          values(name: 'Enabled', type: :dword, data: 0)
          :create
        end
      end
    when 'enable'
      ['AES 128/128', 'AES 256/256'].each do |cipher|
        registry_key cipher do
          key "HKEY_LOCAL_MACHINE\\SYSTEM\\CurrentControlSet\\Control\\SecurityProviders\\SCHANNEL\\Ciphers\\#{cipher}"
          recursive true
          values(name: 'Enabled', type: :dword, data: 4_294_967_295)
          :create
        end
      end
    end
  end
end
